compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensions.cpp

```diff
@@ -815,5 +819,533 @@ DiagnosedSilenceableFailure transform_dialect::ReorderTransposeOp::applyToOne(
+  if (!getEffectsAfter(op->getParentOp(), effects, stopAtBarrier)) return false;
+
+  // If the op is loop-like, collect effects from the leading operations until
+  // we hit a barrier because they can executed after the current operation by
+  // the next iteration of this loop.
+  if (isSequentialLoopLike(op->getParentOp())) {
+    if (isa<gpu::BarrierOp>(op->getBlock()->front())) return true;
+
+    bool exact = collectEffects(&op->getBlock()->front(), effects);
+    return getEffectsAfter(&op->getBlock()->front(), effects,
+                           /*stopAtBarrier=*/true) &&
+           exact;
+  }
+
+  // If the parent operation is not guaranteed to execute its (single-block)
+  // region once, walk the block.
+  bool conservative = false;
+  if (!hasSingleExecutionBody(op->getParentOp()))
+    op->getParentOp()->walk([&](Operation *in) {
+      if (conservative) return WalkResult::interrupt();
+      if (!collectEffects(in, effects)) {
+        conservative = true;
+        return WalkResult::interrupt();
+      }
+      return WalkResult::advance();
+    });
+
+  return !conservative;
+}
+
+/// Looks through known "view-like" ops to find the base memref.
+static Value getBase(Value v) {
+  while (true) {
+    if (auto s = v.getDefiningOp<memref::CastOp>()) {
+      v = s.getSource();
+      continue;
+    }
+    if (auto s = v.getDefiningOp<memref::SubViewOp>()) {
+      v = s.getSource();
+      continue;
+    }
+    if (auto s = v.getDefiningOp<memref::ViewOp>()) {
+      v = s.getSource();
+      continue;
+    }
+    break;
+  }
+  return v;
+}
+
+/// Returns `true` if the value is a defined as a function argument.
+static bool isFunctionArgument(Value v) {
+  auto arg = dyn_cast<BlockArgument>(v);
+  return arg && isa<FunctionOpInterface>(arg.getOwner()->getParentOp());
+}
+
+/// Returns `true` if the value may be captured by any of its users, i.e., if
+/// the user may be storing this value into memory. This makes aliasing analysis
+/// more conservative as it cannot assume the pointer-like value is only passed
+/// around through SSA use-def.
+bool maybeCaptured(Value v) {
+  SmallVector<Value> todo = {v};
+  while (!todo.empty()) {
+    Value v = todo.pop_back_val();
+    for (Operation *user : v.getUsers()) {
+      // A user that is known to only read cannot capture.
+      auto iface = dyn_cast<MemoryEffectOpInterface>(user);
+      if (iface) {
+        SmallVector<MemoryEffects::EffectInstance> effects;
+        iface.getEffects(effects);
+        if (llvm::all_of(effects,
+                         [](const MemoryEffects::EffectInstance &effect) {
+                           return isa<MemoryEffects::Read>(effect.getEffect());
+                         })) {
+          continue;
+        }
+      }
+
+      // When an operation is known to be create an alias, consider if the
+      // source is captured as well.
+      if (auto viewLike = dyn_cast<ViewLikeOpInterface>(user)) {
```

Comment:
collapse, expand, transpose too (I don't think they are defined as "ViewLike" upstream)

Reply:
you could hoist into a separate helper that will later become in interface.